broker/lib/controllers/ServiceBrokerApiController.js

```diff
@@ -160,29 +347,101 @@ class ServiceBrokerApiController extends FabrikBaseController {
       }
       failed(err);
     }
-
     return Promise
       .try(() => req.instance.lastOperation(operation))
       .then(done)
       .catch(AssertionError, failed)
       .catch(ServiceInstanceNotFound, notFound);
   }
 
+  getLastInstanceOperationV2(req, res) {
+    const encodedOp = _.get(req, 'query.operation', undefined);
+    const operation = encodedOp === undefined ? null : utils.decodeBase64(encodedOp);
+    const action = _.capitalize(operation.type);
+    const instanceType = req.instance.constructor.typeDescription;
+    const guid = req.instance.guid;
+
+    function done(result) {
+      const body = _.pick(result, 'state', 'description');
+      // Unlock resource if state is succeeded or failed
+      if (result.state === CONST.OPERATION.SUCCEEDED || result.state === CONST.OPERATION.FAILED) {
+        return lockManager.unlock(req.params.instance_id)
+          .then(() => res.status(CONST.HTTP_STATUS_CODE.OK).send(body));
```

Comment:
If unlock fails, it does not return any response, but throws error. What it means in case of say create, the instance is actually created in background, but CC thinks it is failed or since we don't send any response (failed/success) in this case, CC will call lastOperation again in which case, unlock is called again?
Also, it is needed to add the check like [this](https://github.com/cloudfoundry-incubator/service-fabrik-broker/pull/304/files#diff-03e33ae33c13c68e445aa8cc1048acc9R228)?
Same for [failed](https://github.com/cloudfoundry-incubator/service-fabrik-broker/pull/304/files#diff-03e33ae33c13c68e445aa8cc1048acc9R375) and [gone](https://github.com/cloudfoundry-incubator/service-fabrik-broker/pull/304/files#diff-03e33ae33c13c68e445aa8cc1048acc9R383).

Reply:
@ketakig12 you are right. if unlock fails, we populate error and lastoperation calls keep coming from CF, and as part of lastoperation it will again try to unlock (With 3 retries).
Unlock only fails when apiserver is down, that is unnatural.
Can accommodate suggestions whether to let lastoperation keep coming and wait for unlock to succeed as it doesn't make sense to send 'success' to controller if lock is still present.

Reply:
@amitmalav  - there are two ways to handle when unlock fails:
1. LastOperation must not return back success, but can still return state as 'in-progress' so that CF continues to call back and subsequent deletes could be tried.
2. Even in case of error in unlocking (add 3 retries), if operation has failed we return back operation state as succeeded.
My preference is option #2, because if we go with option #1 for whatsoever reason if unlock keeps failing anyways instance is still locked. However if we go with option #2, retry of unlock should address the failure and all locks should have a TTL and they will be automatically removed post their expiry.

Reply:
Also, my preference is to move out this unlocking from this API layer to the poller when we have all provisioners and pollers in place. Pollers are resilient & in case of failure to unlock, pollers need not be cancelled and they will unlock in their next run.

Reply:
@jagadish-kb Let's discuss the unlock refactoring, will be taken care of as part of that

Reply:
I still see unlock code here and also some `lastOperation` related stuff in `FabrikController` . I am ok with this unlock code being here as it will anyways be removed once unlockPoller is changed to UnlockWatcher which unlocks all types of locks. Either way, please clean up `FabrikController` or this method.